src/common.jl

```diff
@@ -92,8 +92,10 @@ end
 ###########################################################
 
 function sqrt!(a::AbstractArray)
-    @simd for i in eachindex(a)
-        @inbounds a[i] = sqrt(a[i])
+    @inbounds @simd for i in eachindex(a)
+        x = a[i]
+        # > 0 is there to tolerate precision issues
+        a[i] = x > 0 ? sqrt(x) : zero(x)
```

Comment:
I really don't understand why this is needed after my PR but not before it. Without adding this check I get errors because the Mahalanobis distance tries to compute `sqrt(-3.552713678800501e-15)`. This looks like an acceptable approximation to me, but how can my PR have triggered this? Also for safety it would make sense to check that `x` remains small, but without hurting performance too much.

Reply:
I don't think this is the right place to add this. An in-place square root shouldn't truncate here but throw for negative values and the branch will also break vectorization.

Reply:
As I said I don't know why my PR makes this necessary, but without it tests don't pass.

Regarding vectorization, LLVM is able to eliminate the branch and generate SIMD instructions. So this makes absolutely no difference to performance:
```julia
julia> function f(a)
           @inbounds @simd for i in eachindex(a)
               x = a[i]
               a[i] = x > 0 ? sqrt(x) : zero(x)
           end
           a
       end
f (generic function with 1 method)

julia> function g(a)
           @inbounds @simd for i in eachindex(a)
               x = a[i]
               a[i] = sqrt(x)
           end
           a
       end
g (generic function with 1 method)

julia> x = rand(10_000);

julia> @btime f(x);
  20.076 μs (0 allocations: 0 bytes)

julia> @btime g(x);
  20.076 μs (0 allocations: 0 bytes)
```

Overall I think the question is really: is it normal that small negative values are produced? If so, then we need to be prepared to handle them. We could also round them to zero when computing distances. I'm not sure what the tolerance should be, but `eps(1.0)` isn't enough (`sqrt(eps(1.0))` is OK, though).

Reply:
OK, I've found why the tests started failing: I had removed a special method for when a single matrix is passed, which takes advantage from symmetry, which apparently ensures the result is always positive. I'm not sure whether small negative values can legitimately happen, but that's out of the scope of this PR so I have reverted the `sqrt!` change.

Reply:
Actually the problem of small negative values is already known: https://github.com/JuliaStats/Distances.jl/pull/13